[13.0][FIX] account_payment_order: fix get partner_bank_id in account move lines

ValentinVinagre opened the pull request:
Fix get partner_bank_id in account move lines.

**Case:**
The bank account isn't set properly in the payment order lines, because the bank account is only got from account.move.line or the first bank account in the partner_id, but never from the invoice (account.move).

**Solution:**
With this fix, the bank account will be get from:

1. account.move.line
2. account.move
3. first bank account in the partner_id


This fix contain too https://github.com/OCA/bank-payment/issues/701
This fix contain too: fix the attachment file name. https://github.com/OCA/bank-payment/pull/686

ValentinVinagre committed:
[FIX] account_payment_order: fix set partner_bank_id in account move lines
Case:
The bank account isn't set properly in the payment order lines, because the bank account is only got from account.move.line or the first bank account in the partner_id, but never from the invoice (account.move).

Solution:
With this fix, the bank account will be get from account.move when matching the criteria.

ValentinVinagre commented:
Hi @pedrobaeza @carlosdauden @Tardo 
Can you check this fix? is related to 13 MIG account banking sepa credit transfer.
Thanks a lot,

pedrobaeza commented:
I'm not sure about the patch. If `self.partner_bank_id` is not set, the problem is that the transfer from `invoice_partner_bank_id` to the line has not been correct.

ValentinVinagre commented:
The problem I found appears when I run the test. "account.move.line" doesn't have invoice_partner_bank_id set. With this patch this problem is corrected. 
But I don't know if this patch needs to be applied when "account.move.line" is created (invoice validate).
But I agree to change that, if there is any other better way.

pedrobaeza commented:
Is the test failing here in Travis?

ValentinVinagre commented:
The test fails when check the partner_bank_id in the module "account banking sepa credit transfer" if this patch are not applied. https://github.com/OCA/bank-payment/pull/699
Here is the link:
https://travis-ci.org/github/OCA/bank-payment/jobs/678673610

Header explains the fail.

HaraldPanten commented:
Hi @ValentinVinagre ,

Is it OK for you, to add https://github.com/OCA/bank-payment/issues/701 in this PR, please?

Thank you in advance! 👍

ValentinVinagre commented:
Hi @HaraldPanten ,
no problem to fix, but I don't know if I need open other PR for resolve the https://github.com/OCA/bank-payment/issues/701 or I can add the solution in this fix, @pedrobaeza how should I proceed?

Thanks,

ValentinVinagre commented:
This fix contain too https://github.com/OCA/bank-payment/issues/701

ValentinVinagre commented:
Hi @pedrobaeza ,
Have you been able to review this problem?

Thank you,

HaraldPanten reviewed (approved):
Functional review 👍 

IMO, both fixes are working fine.

HaraldPanten reviewed (approved):
Tested again with the new fix in local environment.

Functional review OK. 👍

HaraldPanten commented:
Hi @pedrobaeza could you have a look? This fix is blocking several dependencies.

Thanks in advance!

pedrobaeza commented:
This has 3 things, and I don't agree with the main one: we shouldn't change the way payment lines are prepared, but see why move line is not having propagated the bank account from invoice (account.move) to the AR/AP account.move.line. This is the patch to do.

ValentinVinagre commented:
I will check to see if I can fix it in another way.

ValentinVinagre commented:
Hi @pedrobaeza ,
I see in v.12 that fields are set in https://github.com/OCA/bank-payment/blob/d5f9c89399a4c8fbd55b876e7d26039c6f80c17b/account_payment_order/models/account_invoice.py#L51 but this functionality has not been implemented in the migration to v.13.

I added this field when the user posts the "invoices". Is this solution more suitable?

pedrobaeza commented:
Please don't do it on post, but on account.move.line creation, where there are hooks for that, and I would say that the same hook is used for updating values. You can put also a computed writable field if you prefer it. See for example my refactoring using this new feature in this module, that saves tons of code:

https://github.com/OCA/commission/pull/237/commits/081678a50f98c7a75ffe13809e4875558e60c7d0

ValentinVinagre commented:
Hi @pedrobaeza ,
done.

StefanRijnhart commented on account_payment_order/views/ir_attachment.xml:
```diff
@@ -12,12 +12,7 @@
                         <field name="name" />
                     </h1>
                     <group name="main">
-                        <field
```
Can I ask if this alternative attachment view is in need of a value for the field `priority` so that it does not compete with the main attachment form view?

ValentinVinagre replied:
it's possible but this FIX have 3 fix's and I don't want to add more. If the view haver any problem open other fix please.

StefanRijnhart replied:
Yes, I was wondering about adding this second fix but given that you are modifying this view you could either split off all view fixes into a separate PR or add the priority + view selection in this PR.

ValentinVinagre replied:
If you tell me I should change, I can add the changes in this solution 👍

StefanRijnhart replied:
Thanks, but you are right about combining too many different fixes in the same PR. I've proposed this priority thing here: https://github.com/OCA/bank-payment/pull/712/files

StefanRijnhart commented on account_payment_order/models/account_move_line.py:
```diff
@@ -23,6 +26,25 @@ class AccountMoveLine(models.Model):
     )
 
+    @api.depends(
+        "move_id", "move_id.invoice_partner_bank_id", "move_id.payment_mode_id"
```
Isn't the partner_bank_id redundant now that Odoo added an invoice_partner_bank_id on the move level?

ValentinVinagre replied:
I don't think so, I explain one situation:
You have to pay your employees, so you will create an "account move" to generate this movement. Once you have to pay it, you want to associate the "account.move.line" to a payment order, so you have to be able to associate a "bank account" to an "account.move.line".

StefanRijnhart replied:
Yes, good point.

StefanRijnhart commented on account_payment_order/models/account_move_line.py:
```diff
@@ -12,6 +12,9 @@ class AccountMoveLine(models.Model):
     partner_bank_id = fields.Many2one(
         comodel_name="res.partner.bank",
         string="Partner Bank Account",
+        compute="_compute_partner_bank_id",
```
I think making this a computed field could maybe lead to unexpected results. Would you consider, instead of updating the move line's partner_bank_id directly to change the logic here: https://github.com/OCA/bank-payment/blob/13.0/account_payment_order/models/account_move_line.py#L58 and fetch either the move line's partner_bank_id when it's set, or the move's invoice_partner_bank_id, or else any of the partner's banks?

ValentinVinagre replied:
That solution is the first solution he presents, but he wants to solve the reason why he did not transfer the bank account, therefore it was not well received. Now the third solution is presented, you can read the solution conversation where this solution is recommended.

StefanRijnhart replied:
Oh sh*t that was your first proposal but @pedrobaeza did not like it. Well, then I will refrain from reviewing this PR further (refer to deleted commit https://github.com/OCA/bank-payment/commit/b7b08ac44672bf51968e974998d5e6aca2877c22)

ValentinVinagre replied:
Haha no problem 👍

HaraldPanten reviewed (approved):
👍 IMO. What do you think @StefanRijnhart ?

StefanRijnhart commented:
> +1 IMO. What do you think @StefanRijnhart ?

Like I said, I like the earlier version better. Seems to me that the move's invoice_partner_bank_id serves a different purpose than the move line's partner_bank_id and trying to keep them synced might lead to unexpected behaviour. But I'm refraining from review so that I'm not blocking the merge of this.

ValentinVinagre commented:
Hi @pedrobaeza ,
Could you review? I change the "partner_bank_id" field by adding a computed method.

HaraldPanten commented:
Hi @pedrobaeza we applied the changes that you suggested. Is there anything else we need to do before merging this FIX? It's blocking other PR's. Could you have a look?

Thanks!

pedrobaeza commented on account_payment_order/models/account_move_line.py:
```diff
@@ -23,6 +26,25 @@ class AccountMoveLine(models.Model):
+    )
+    def _compute_partner_bank_id(self):
+        for ml in self:
+            res = None
```
Better to avoid possible side effects using a falsy but not expected value

```suggestion
            res = False
```

pedrobaeza commented on account_payment_order/models/account_move_line.py:
```diff
@@ -23,6 +26,25 @@ class AccountMoveLine(models.Model):
+                )
+            ):
+                res = ml.move_id.invoice_partner_bank_id.id
+            ml.partner_bank_id = res
```
Anyway, I think you mustn't assign always the bank account, but only when matching the criteria. If not, you are emptying the field in the move line is linked to a payment line for example.

ValentinVinagre replied:
done.

pedrobaeza reviewed (changes requested):
And the other unrelated changes?

ValentinVinagre commented:
@pedrobaeza changes done.

pedrobaeza commented:
And what about the other unrelated changes? Why are they here?

ValentinVinagre commented:
The unrelated changes are: 
This fix contain too #701 
This fix contain too: fix the attachment file name. #686

I put all the changes in the same PR because they are very small fixes.
If you prefer I can create other 2 PR's with this fixes, and only leave the 'partner_bank_id' fix in this PR.

pedrobaeza commented:
Well, they can go in the same PR being small, but different commits explaining each fix. Current commit message is:

- Very short. Please include the use case and description you have put in the PR main comment. When browsing commit history, you don't see the PRs where they were commented.
- Only referring to the first fix, not the others. But as said, it's preferable in this case to have a commit for each fix.

ValentinVinagre committed:
[FIX] account_payment_order: fix payment orders from invoice tree view
Case: User can't add invoices to payment/debit orders from the tree view but it's still possible to add them one by one from the form view.
- Remove binding_views="form": by default are "tree & form".
With this fix, users can add payment/debit orders from the invoice tree view

ValentinVinagre committed:
[FIX] account_payment_order: fix the attachment file name & dowload.
Case: After the configuration of a debit order and generating the .xml file, user can't download it from the next step screen using the download link.
This fix makes the .xml file downloadable from the .xml file name as it was in older versions

ValentinVinagre commented:
> Well, they can go in the same PR being small, but different commits explaining each fix. Current commit message is:
> 
> * Very short. Please include the use case and description you have put in the PR main comment. When browsing commit history, you don't see the PRs where they were commented.
> * Only referring to the first fix, not the others. But as said, it's preferable in this case to have a commit for each fix.

Done in 3 commits and each one has been commented.

pedrobaeza reviewed (approved):
Thanks for the efforts. Now it's all homogenized and well conducted.

pedrobaeza commented:
/ocabot merge minor
